frame/pools/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+	/// Merge the oldest unbonding pool with an era into the general unbond pool with no associated
+	/// era.
+	fn maybe_merge_pools(mut self, current_era: EraIndex) -> Self {
+		if current_era < T::MaxUnbonding::get().into() {
+			// For the first `T::MAX_UNBONDING` eras of the chain we don't need to do anything.
+			// I.E. if `MAX_UNBONDING` is 5 and we are in era 4 we can add a pool for this era and
+			// have exactly `MAX_UNBONDING` pools.
+			return self
+		}
+
+		//  I.E. if `MAX_UNBONDING` is 5 and current era is 10, we only want to retain pools 6..=10.
+		let oldest_era_to_keep = current_era - T::MaxUnbonding::get().saturating_add(1);
+
+		let eras_to_remove: Vec<_> =
+			self.with_era.keys().cloned().filter(|era| *era < oldest_era_to_keep).collect();
+		for era in eras_to_remove {
+			if let Some(p) = self.with_era.remove(&era) {
+				self.no_era.points.saturating_add(p.points);
+				self.no_era.balance.saturating_add(p.balance);
+			} else {
+				// the world is broken
+			}
+		}
+
+		self
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::{ensure_signed, pallet_prelude::*};
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	#[pallet::generate_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Weight information for extrinsics in this pallet.
+		// type WeightInfo: weights::WeightInfo;
+
+		/// The nominating balance.
+		type Currency: Currency<Self::AccountId>;
+
+		// Infallible method for converting `Currency::Balance` to `u128`.
+		type BalanceToU128: Convert<BalanceOf<Self>, u128>;
+
+		// Infallible method for converting `u128` to `Currency::Balance`.
+		type U128ToBalance: Convert<u128, BalanceOf<Self>>;
+
+		/// The interface for nominating.
+		type StakingInterface: StakingInterface<
+			Balance = BalanceOf<Self>,
+			AccountId = Self::AccountId,
+			LookupSource = <Self::Lookup as StaticLookup>::Source,
+		>;
+
+		/// The maximum amount of eras an unbonding pool can exist prior to being merged with the
+		/// `no_era	 pool. This should at least be greater then the `UnbondingDuration` for staking
+		/// so delegator have a chance to withdraw unbonded before their pool gets merged with the
+		/// `no_era` pool. This *must* at least be greater then the slash deffer duration.
+		type MaxUnbonding: Get<u32>;
+	}
+
+	/// Active delegators.
+	#[pallet::storage]
+	pub(crate) type Delegators<T: Config> =
+		CountedStorageMap<_, Twox64Concat, T::AccountId, Delegator<T>>;
+
+	/// `PoolId` lookup from the pool's `AccountId`. Useful for pool lookup from the slashing
+	/// system.
+	#[pallet::storage]
+	pub(crate) type PoolIds<T: Config> = CountedStorageMap<_, Twox64Concat, T::AccountId, PoolId>;
+
+	/// Bonded pools.
+	#[pallet::storage]
+	pub(crate) type PrimaryPools<T: Config> =
```

Comment:
Kinda torn between the terminology: is bonded pool and primary pool the same thing? if yes, why do we need to call them differently?

Reply:
Basically:

PrimaryPool => active bonded
SubPools => unbond{ing,ed} pools

Early on I was tripping myself up on boonded,unbonded,unbonding because they look very similar so I decided to try out this alternative naming scheme. Totally open to changing - I just wanted names that where both meaningful not easy to confuse with each other

Reply:
What you say about sub-pools makes sense, I just think avoiding the prefix `primary` is better and we can stick to BondedPool. If you are having issue with the fact that the struct and the `#[pallet:storage]` have the same name, I recommend (and already doing) calling it like `#[palle::storage] pub type BondedPoolStorage = StorageValue<_, _, BondedPool>` or something like that.